jujugui/static/gui/src/app/components/user-profile/model-list/model-list.js

```diff
@@ -108,6 +110,134 @@ YUI.add('user-profile-model-list', function() {
+    },
+
+    /**
+      Generate the confirmation for destroying a model.
+
+      @method _displayConfirmation
+      @return {Object} the confirmation component.
+    */
+    _generateConfirmation: function() {
+      const model = this.state.destroyModel;
+      const addNotification = this.props.addNotification;
+      if (!model) {
+        return;
+      }
+      if (model.name === 'controller') {
+        addNotification({
+          title: 'Cannot destroy model',
+          message: 'The controller model cannot be destroyed. ',
+          level: 'error'
+        });
+        return;
+      }
+      const buttons = [{
+        title: 'Cancel',
+        action: this._displayConfirmation.bind(this, null),
+        type: 'base'
+      }, {
+        title: 'Destroy',
+        action: this._destroyModel,
+        type: 'destructive'
+      }];
+      const name = model.name.split('/')[1];
+      const message = `Are you sure you want to destroy ${name}?`
+        + ' All the applications and units included in the model will be'
+        + ' destroyed. This action cannot be undone.';
+      return (
+        <juju.components.ConfirmationPopup
+          buttons={buttons}
+          message={message}
+          title="Destroy model" />);
+    },
+
+    /**
+      Makes a request of JIMM or JES to delete a selected model.
+
+      @method _destroyModel
+    */
+    _destroyModel: function() {
+      const model = this.state.destroyModel;
+      const uuid = model.uuid;
+      // Hide the confirmation popup.
+      this._displayConfirmation(null);
+      const xhr = this.props.destroyModels([uuid], this._destroyModelCallback);
+      this.xhrs.push(xhr);
+    },
+
+    /**
+      Callback for the JIMM and JES delete model call.
+
+      @method _destroyModelCallback
+      @param {String} err The error from the request, or null.
+      @param {Array} modelList The list of models.
+    */
+    _destroyModelCallback: function(err, results) {
+      const addNotification = this.props.addNotification;
+      // Handle global errors.
+      if (err) {
+        addNotification({
+          title: 'Model destruction failed',
+          message: 'The model failed to be destroyed: ' + err,
+          level: 'error'
+        });
+        return;
+      }
+      // Check for bad data.
+      let uuids;
+      try {
+        uuids = Object.keys(results);
+      } catch (e) {
+        addNotification({
+          title: 'Model destruction failed',
+          message: 'Bad data returned from model destruction.',
+          level: 'error'
+        });
+        return;
+      }
+      if (!uuids || !uuids.length || uuids.length !== 1) {
+        addNotification({
+          title: 'Model destruction failed',
+          message: 'Bad data returned from model destruction.',
+          level: 'error'
+        });
+        return;
+      }
+      const uuid = uuids[0];
+      // Check for model-specific errors.
+      const uuidError = results[uuid];
```

Comment:
This is the proper error handling expected here (thank you!): this and the global error handling done at the beginning of this callback. The other stuff should be removed from here and handled at the API layer level IMHO. What do you think?